examples/vae_text/vae_train.py

```diff
@@ -0,0 +1,463 @@
+                                      device=device)
+    val_data = tx.data.MonoTextData(config.val_data_hparams,
+                                    device=device)
+    test_data = tx.data.MonoTextData(config.test_data_hparams,
+                                     device=device)
+
+    iterator = tx.data.TrainTestDataIterator(train=train_data,
+                                             val=val_data,
+                                             test=test_data)
+
+    opt_vars = {
+        'learning_rate': config.lr_decay_hparams["init_lr"],
+        'best_valid_nll': 1e100,
+        'steps_not_improved': 0,
+        'kl_weight': config.kl_anneal_hparams["start"]
+    }
+
+    decay_cnt = 0
+    max_decay = config.lr_decay_hparams["max_decay"]
+    decay_factor = config.lr_decay_hparams["decay_factor"]
+    decay_ts = config.lr_decay_hparams["threshold"]
+
+    save_dir = f"./models/{config.dataset}"
+
+    if not os.path.exists(save_dir):
+        os.makedirs(save_dir)
+
+    suffix = f"{config.dataset}_{config.decoder_type}Decoder.ckpt"
+
+    save_path = os.path.join(save_dir, suffix)
+
+    # KL term annealing rate
+    anneal_r = 1.0 / (config.kl_anneal_hparams["warm_up"] *
+        (train_data.__len__() / config.batch_size))
+
+    model = Vae(train_data)
+    model.to(device)
+
+    optimizer = tx.core.get_optimizer(params=model.parameters(),
+                                    hparams=config.opt_hparams)
+
+    def _run_epoch(epoch, mode, display=10):
+        if mode == 'train':
+            iterator.switch_to_train_data()
+        elif mode == 'valid':
+            iterator.switch_to_val_data()
+        elif mode == 'test':
+            iterator.switch_to_test_data()
+        if mode == 'train':
+            model.train()
+        else:
+            model.eval()
+        step = 0
+        start_time = time.time()
+        num_words = num_sents = 0
+        nll_ = 0.
+        kl_loss_ = rc_loss_ = 0.
+
+        for batch in iterator:
+            if mode == 'train':
+                opt_vars["kl_weight"] = min(
+                    1.0, opt_vars["kl_weight"] + anneal_r)
+
+                kl_weight_ = opt_vars["kl_weight"]
+            else:
+                kl_weight_ = 1.0
+
+            mode = mode
+
+            if mode == "train":
+                adjust_learning_rate(optimizer, opt_vars["learning_rate"])
+
+            fetches_ = model(batch, kl_weight_)
+            if mode == "train":
+                fetches_["nll"].backward()
+                optimizer.step()
+                optimizer.zero_grad()
+
+            batch_size_ = len(fetches_["lengths"])
+            num_sents += batch_size_
+            num_words += sum(fetches_["lengths"])
+            nll_ += fetches_["nll"].item() * batch_size_
+            kl_loss_ += fetches_["kl_loss"] * batch_size_
+            rc_loss_ += fetches_["rc_loss"] * batch_size_
+
+            if step % display == 0 and mode == 'train':
+                nll = nll_ / float(num_sents)
```

Comment:
Just as a side note: no need to call `float` here. Division by a single slash (`/`) always returns floating point numbers. Double slash (`//`) is integer division.

Reply:
Corrected.